src/dxil/wrapper.rs

```diff
@@ -93,3 +95,353 @@ impl DxilContainerReflection {
         );
     }
 }
+
+#[derive(Debug)]
+pub struct D3D12LibraryReflection {
+    inner: ComPtr<ID3D12LibraryReflection>,
+}
+
+impl D3D12LibraryReflection {
+    fn new(inner: ComPtr<ID3D12LibraryReflection>) -> Self {
+        Self { inner }
+    }
+
+    pub fn get_desc(&self) -> Result<d3d12shader::D3D12_LIBRARY_DESC, HassleError> {
+        let mut desc = d3d12shader::D3D12_LIBRARY_DESC::default();
+        return_hr_wrapped!(unsafe { self.inner.get_desc(&mut desc as *mut _) }, desc);
+    }
+
+    pub fn get_function_by_index(&self, function_index: i32) -> D3D12FunctionReflection {
+        unsafe { D3D12FunctionReflection::new(self.inner.get_function_by_index(function_index)) }
+    }
+}
+
+#[derive(Debug)]
+pub struct D3D12FunctionParameterReflection {
+    inner: ComPtr<ID3D12FunctionParameterReflection>,
+}
+
+impl D3D12FunctionParameterReflection {
+    fn new(inner: ComPtr<ID3D12FunctionParameterReflection>) -> Self {
+        Self { inner }
+    }
+
+    pub fn get_desc(&self) -> Result<d3d12shader::D3D12_PARAMETER_DESC, HassleError> {
+        let mut desc = d3d12shader::D3D12_PARAMETER_DESC::default();
+        return_hr_wrapped!(unsafe { self.inner.get_desc(&mut desc as *mut _) }, desc);
+    }
+}
+
+#[derive(Debug)]
+pub struct D3D12FunctionReflection {
+    inner: ComPtr<ID3D12FunctionReflection>,
+}
+
+impl D3D12FunctionReflection {
+    fn new(inner: ComPtr<ID3D12FunctionReflection>) -> Self {
+        Self { inner }
+    }
+
+    pub fn get_constant_buffer_by_index(
+        &self,
+        buffer_index: u32,
+    ) -> D3D12ShaderReflectionConstantBuffer {
+        unsafe {
+            D3D12ShaderReflectionConstantBuffer::new(
+                self.inner.get_constant_buffer_by_index(buffer_index),
+            )
+        }
+    }
+
+    pub fn get_constant_buffer_by_name(&self, name: LPCSTR) -> D3D12ShaderReflectionConstantBuffer {
```

Comment:
De safe bindings mogen ook geen raw types zoals een `LPCSTR` (`*const CHAR`) naar buiten lekken. Dit moet een `&str` zijn, hieronder roep je dan gewoon `name.as_ptr()` aan.

Reply:
EDIT: Dat is niet helemaal waar, omdat de C strings null-terminated moeten zijn, terwijl Rust in de fat pointer opslaat hoe lang de string slice is (`.as_ptr()` returnt een string zonder null-termination). De enige optie is een `CStr` maken met explciet `'\0'`-character aan het einde (via `from_bytes_with_nul`) voor true zero-copy of een nieuwe allocation doen waarin deze string met null-character zit via `CString::new` zoals in de rest van de code.